### ISSUE TITLE
Constrained TypeVar failing when shared between two classes

### ISSUE THREAD

**Issue by ian-r-rose** [microsoft/pyright#2150]
**Describe the bug**

I'm getting some unexpected behavior when two classes are using the same constrained `TypeVar`, where it works with some types but not with `Literal[True]` or `Literal[False]`.

**To Reproduce**

The following classes `A` and `B` are both generic with the same typevar, and instances of `B` should have attributes of type `A` constrained to the same type. This works as expected when the types are literal strings:
```python
from typing import TypeVar, Generic, Literal

XOrY = TypeVar("XOrY", Literal["X"], Literal["Y"])

class A(Generic[XOrY]):
    pass

class B(Generic[XOrY]):
    def __init__(self, a: A[XOrY]):
        self.a = a
```

But when I change the constrained types to `Literal[True]`, `Literal[False]` I get an error:
```python
from typing import TypeVar, Generic, Literal

XOrY = TypeVar("XOrY", Literal[True], Literal[False])

class A(Generic[XOrY]):
    pass

class B(Generic[XOrY]):
    def __init__(self, a: A[XOrY]):
        self.a = a
```
```python
 11:29 - error: Type "XOrY@B" cannot be assigned to type variable "XOrY@A"
    Type "bool*" is incompatible with constrained type variable "XOrY@A" (reportGeneralTypeIssues)
```
Interestingly, if I add `bool` to the list of constrained types, then the above passes, so perhaps there is some additional logic behind the scenes specific to boolean literals that I am misunderstanding?

**Expected behavior**

I expected the second snippet to pass in the same way the first one did.

**VS Code extension or command-line**

Pyright command line @1.1.159. As far as I can tell, the above used to work, but then started failing in 1.1.145

**Comment by erictraut** [comment]
Thanks for the bug report. The type analyzer contains some code that coalesces a union of `Literal[False]` and `Literal[True]` into `bool`. It was inappropriately applying this trick in one place, which led to the bug. This will be fixed in the next release.

**Comment by erictraut** [comment]
This is fixed in pyright 1.1.160, which I just published. It will also be included in the next release of pylance.

**Comment by ian-r-rose** [comment]
Thanks!
